confidential-datahub/kms/src/plugins/aliyun/client.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::BTreeMap;
+
+use async_trait::async_trait;
+use base64::engine::general_purpose::STANDARD;
+use base64::Engine;
+use chrono::Utc;
+use log::error;
+use prost::Message;
+use reqwest::Certificate;
+use reqwest::{header::HeaderMap, ClientBuilder};
+use serde_json::Value;
+use sha2::{Digest, Sha256};
+use tokio::fs;
+
+use crate::plugins::aliyun::client::dkms_api::{DecryptRequest, EncryptRequest};
+use crate::{Decryptor, Encryptor, OperationContext, ProviderSetting};
+use crate::{Error, Result};
+
+use super::annotations::{CipherContext, KmsContext};
+use super::credential::Credential;
+
+pub mod dkms_api {
+    tonic::include_proto!("dkms_api");
+}
+
+pub struct AliyunKmsClient {
+    http_client: reqwest::Client,
+    credential: Credential,
+    kms_instance_id: String,
+    endpoint: String,
+}
+
+const IN_POD_DEFAULT_KEY_PATH: &str = "/run/kms-credential/aliyun";
+
+impl AliyunKmsClient {
+    fn read_kms_instance_cert(cert_pem: &[u8]) -> Result<Certificate> {
+        let kms_instance_ca_cert = Certificate::from_pem(cert_pem)
+            .map_err(|e| Error::AliyunKmsError(format!("read kms instance ca cert failed: {e}")))?;
+        Ok(kms_instance_ca_cert)
+    }
+
+    pub fn new(
+        client_key: &str,
+        kms_instance_id: &str,
+        pswd: &str,
+        cert_pem: &str,
+    ) -> Result<Self> {
+        let credential = Credential::new(client_key, pswd).map_err(|e| {
+            Error::AliyunKmsError(format!("create credential of the kms instance failed: {e}"))
+        })?;
+        let endpoint = format!("https://{kms_instance_id}.cryptoservice.kms.aliyuncs.com");
+        let cert = Self::read_kms_instance_cert(cert_pem.as_bytes())?;
+        let http_client = ClientBuilder::new()
+            .add_root_certificate(cert)
+            .build()
+            .map_err(|e| Error::AliyunKmsError(format!("build http client failed: {e}")))?;
+
+        Ok(Self {
+            http_client,
+            kms_instance_id: kms_instance_id.to_string(),
+            credential,
+            endpoint,
+        })
+    }
+
+    /// This new function is used by a in-pod client. The side-effect is to read the
```

Comment:
What do you mean by an "in-pod client?"

Reply:
The client definitions in `kms` crate are generic things. They can be used by a client tool to generate sealed secret (user side). Also they can be used by CDH to unseal secrets (so-called in-pod client, as it works in the pod)

Reply:
When you say "in-pod" do you mean in-guest? To me in-pod sounds like it is running in a container.

Reply:
Oh. Ok, let me use `in-guest` then. Fixed this in the code notes.